crates/server/tests/config_tests.rs

```diff
@@ -273,3 +279,225 @@ define_test_config!(openssl_dnssec);
+                    let mut output = self.original.clone();
+                    output[*key] = Value::Table(table);
+                    return Some(output);
+                } else {
+                    self.nested_table_mutator = None;
+                }
+            }
+            if let Some((key, iter)) = self.nested_array_mutator.as_mut() {
+                if let Some(array) = iter.next() {
+                    let mut output = self.original.clone();
+                    output[*key] = Value::Array(array);
+                    return Some(output);
+                } else {
+                    self.nested_array_mutator = None;
+                }
+            }
+            if let Some(index) = self.index_iter.next() {
+                match self.original.get(index).unwrap() {
+                    Value::String(_)
+                    | Value::Integer(_)
+                    | Value::Float(_)
+                    | Value::Boolean(_)
+                    | Value::Datetime(_) => {}
+                    Value::Array(array) => {
+                        self.nested_array_mutator =
+                            Some((index, Box::new(ArrayMutator::new(array))));
+                    }
+                    Value::Table(table) => {
+                        self.nested_table_mutator =
+                            Some((index, Box::new(TableMutator::new(table))));
+                    }
+                }
+            } else {
+                return None;
+            }
+        }
+    }
+}
+
+/// Check that unknown fields in configuration files are rejected. This uses each example
+/// configuration file as a seed, and tries adding invalid fields to each table.
+#[test]
+fn test_reject_unknown_fields() {
+    let test_configs_dir =
+        PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../../tests/test-data/test_configs");
+    for result in read_dir(test_configs_dir).unwrap() {
+        let entry = result.unwrap();
+        let file_name = entry.file_name().into_string().unwrap();
+        if !file_name.ends_with(".toml") {
+            continue;
+        }
+        println!("seed file: {file_name}");
+
+        let mut file = File::open(entry.path()).unwrap();
+        let mut contents = String::new();
+        file.read_to_string(&mut contents).unwrap();
+        let value = toml::from_str::<toml::Value>(&contents).unwrap();
+        let config_table = value.as_table().unwrap();
+
+        // Skip over configs that can't be read with the current set of features.
+        #[allow(unused_mut)]
+        let mut skip = false;
+        #[cfg(not(feature = "dnssec"))]
+        if config_table.contains_key("tls_cert") {
+            println!("skipping due to tls_cert setting");
+            skip = true;
+        }
+        if let Some(zones) = config_table.get("zones") {
+            let zones = zones.as_array().unwrap();
+            for zone in zones {
+                if let Some(stores) = zone.get("stores") {
```

Comment:
Nit: suggest doing an `let Some(stores) = zones.get("stores") else { continue }` here to reduce rightward drift.